### PR TITLE
Add .NET Core support for OpenWhisk

### PR DESCRIPTION
**Pull request by shawnallen85**

### CODE REVIEW

File: core/dotnet2.1/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
+FROM microsoft/dotnet:2.1-sdk-alpine AS build
```
Comment (rabbah) [inline core/dotnet2.1/Dockerfile:18]: needs license header  see https://github.com/apache/incubator-openwhisk-runtime-nodejs/blob/master/core/nodejs6Action/Dockerfile

File: core/dotnet2.1/build-run.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+# Removes all previously built instances.
+
+docker build -t openwhisk-runtime-dotnet .
+
+echo ""
+echo "  ---- RUNNING ---- "
+echo ""
+
+docker run --rm -it -p 8080:8080 openwhisk-runtime-dotnet
```
Comment (rabbah) [inline core/dotnet2.1/build-run.sh]: you don't need this in that the gradle command will rebuild the container for you.

File: core/dotnet2.1/proxy/Apache.OpenWhisk.Runtime.Common/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace Apache.OpenWhisk.Runtime.Common
+{
+    public class Startup
+    {
+        public static void WriteLogMarkers()
+        {
+            Console.WriteLine("XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX");
+            Console.Error.WriteLine("XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX");
+            //await Console.Error.FlushAsync();
```
Comment (rabbah) [inline core/dotnet2.1/proxy/Apache.OpenWhisk.Runtime.Common/Startup.cs]: remove commented line?

File: core/dotnet2.1/proxy/Apache.OpenWhisk.Runtime.Common/Run.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    );
+                }
+            }
+            finally
+            {
+                Startup.WriteLogMarkers();
+            }
+        }
+    }
+}
```
Comment (rabbah) [inline core/dotnet2.1/proxy/Apache.OpenWhisk.Runtime.Common/Run.cs]: missing line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/AltEcho.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    public class AltEcho
+    {
+        public JObject Main(JObject args)
+        {
+            Console.WriteLine("hello stdout");
+            Console.Error.WriteLine("hello stderr");
+            return (args);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/AltEcho.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Echo.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+namespace Apache.OpenWhisk.Tests.Dotnet
+{
+    public class Echo
+    {
+        public JObject Main(JObject args)
+        {
+            return (args);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Echo.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Environment.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            message.Add("api_host", new JValue(System.Environment.GetEnvironmentVariable("__OW_API_HOST")));
+            message.Add("api_key", new JValue(System.Environment.GetEnvironmentVariable("__OW_API_KEY")));
+            message.Add("namespace", new JValue(System.Environment.GetEnvironmentVariable("__OW_NAMESPACE")));
+            message.Add("action_name", new JValue(System.Environment.GetEnvironmentVariable("__OW_ACTION_NAME")));
+            message.Add("activation_id", new JValue(System.Environment.GetEnvironmentVariable("__OW_ACTIVATION_ID")));
+            message.Add("deadline", new JValue(System.Environment.GetEnvironmentVariable("__OW_DEADLINE")));
+            return (message);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Environment.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Error.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    public class Error
+    {
+        public JObject Main(JObject args)
+        {
+            JObject message = new JObject();
+            message.Add("error", new JValue("This action is unhappy."));
+            return (message);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Error.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Exception.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+namespace Apache.OpenWhisk.Tests.Dotnet
+{
+    public class Exception
+    {
+        public JObject Main(JObject args)
+        {
+            throw (new System.Exception("noooooooo"));
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Exception.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Unicode.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            string delimiter = args["delimiter"].ToString();
+            JObject message = new JObject();
+            string output = $"{delimiter} ☃ {delimiter}";
+            message.Add("winter", new JValue(output));
+            Console.WriteLine(output);
+            return (message);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Unicode.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/NonEmptyConstructor.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        public NonEmptyConstructor(string value)
+        {
+            System.Console.WriteLine(value);
+        }
+        public JObject Main(JObject args)
+        {
+            return (args);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/NonEmptyConstructor.cs]: line return.

File: tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Nuller.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+namespace Apache.OpenWhisk.Tests.Dotnet
+{
+    public class Nuller
+    {
+        public JObject Main(JObject args)
+        {
+            return (null);
+        }
+    }
+}
```
Comment (rabbah) [inline tests/dotnet/Apache.OpenWhisk.Tests.Dotnet/Nuller.cs]: line return.

File: tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+        e shouldBe empty
+    })
+  }
+
+  it should "fails on invalid assembly reference" in {
+    val (out, err) = withActionContainer() { c =>
+      val (initCode, initRes) = c.init(initPayload(functionb64, "Apache.OpenWhisk.Tests.Dotnet.DoesntExist::Apache.OpenWhisk.Tests.Dotnet.Echo::Main"))
+      initCode should be(502)
+
+      initRes shouldBe defined
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala:137]: redundant with next statement.

File: tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+      }
+    }
+  }
+
+  it should "fails on invalid type reference" in {
+    val (out, err) = withActionContainer() { c =>
+      val (initCode, initRes) = c.init(initPayload(functionb64, "Apache.OpenWhisk.Tests.Dotnet::Apache.OpenWhisk.Tests.Dotnet.FakeType::Main"))
+      initCode should be(502)
+
+      initRes shouldBe defined
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala]: `       initRes shouldBe defined ` not necessary next statement checks result is Some...

File: tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+      }
+    }
+  }
+
+  it should "fails on invalid method reference" in {
+    val (out, err) = withActionContainer() { c =>
+      val (initCode, initRes) = c.init(initPayload(functionb64, "Apache.OpenWhisk.Tests.Dotnet::Apache.OpenWhisk.Tests.Dotnet.Echo::FakeMethod"))
+      initCode should be(502)
+
+      initRes shouldBe defined
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala]: 163 not necessary, see above.

File: tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+      }
+    }
+  }
+
+  it should "fails on type with no empty constructor" in {
+    val (out, err) = withActionContainer() { c =>
+      val (initCode, initRes) = c.init(initPayload(functionb64, "Apache.OpenWhisk.Tests.Dotnet::Apache.OpenWhisk.Tests.Dotnet.NonEmptyConstructor::Main"))
+      initCode should be(502)
+
+      initRes shouldBe defined
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala]: 176 not needed.

File: tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+      }
+    }
+  }
+
+  it should "validate main string format 1" in {
+    val (out, err) = withActionContainer() { c =>
+      val (initCode, initRes) = c.init(initPayload(functionb64, "Apache.OpenWhisk.Tests.Dotnet"))
+      initCode should not be (200)
+
+      initRes shouldBe defined
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala]: 189 not needed.

File: tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+      }
+    }
+  }
+
+  it should "validate main string format 2" in {
+    val (out, err) = withActionContainer() { c =>
+      val (initCode, initRes) = c.init(initPayload(functionb64, "Apache.OpenWhisk.Tests.Dotnet::Apache.OpenWhisk.Tests.Dotnet.Echo"))
+      initCode should not be (200)
+
+      initRes shouldBe defined
```
Comment (rabbah) [inline tests/src/test/scala/actionContainers/DotNetActionContainerTests.scala]: 202 not needed.

**Review by rabbah — APPROVED** [review]

**Comment by csantanapr** [comment]
thanks @rabbah for taking a look at the PR 👍

**Pull request merged**